ddtrace/contrib/internal/google_generativeai/patch.py

```diff
@@ -0,0 +1,113 @@
+from ddtrace.contrib.internal.google_generativeai._utils import tag_request
+from ddtrace.contrib.internal.google_generativeai._utils import tag_response
+from ddtrace.contrib.trace_utils import unwrap
+from ddtrace.contrib.trace_utils import with_traced_module
+from ddtrace.contrib.trace_utils import wrap
+from ddtrace.llmobs._integrations import GeminiIntegration
+from ddtrace.pin import Pin
+
+
+config._add(
+    "genai",
+    {
+        "span_prompt_completion_sample_rate": float(
+            os.getenv("DD_GOOGLE_GENAI_SPAN_PROMPT_COMPLETION_SAMPLE_RATE", 1.0)
+        ),
+        "span_char_limit": int(os.getenv("DD_GOOGLE_GENAI_SPAN_CHAR_LIMIT", 128)),
+    },
+)
+
+
+def get_version():
+    # type: () -> str
+    return getattr(genai, "__version__", "")
+
+
+@with_traced_module
+def traced_generate(genai, pin, func, instance, args, kwargs):
+    integration = genai._datadog_integration
+    stream = kwargs.get("stream", False)
+    generations = None
+    span = integration.trace(
+        pin,
+        "%s.%s" % (instance.__class__.__name__, func.__name__),
+        provider="google",
+    )
+    try:
+        tag_request(span, integration, instance, args, kwargs)
+        generations = func(*args, **kwargs)
+        api_key = _extract_api_key(instance)
+        if api_key:
+            span.set_tag("genai.request.api_key", "...{}".format(api_key[-4:]))
+        if stream:
+            return TracedGenerateContentResponse(generations, instance, integration, span, args, kwargs)
+        tag_response(span, generations, integration, instance)
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        raise
+    finally:
+        # streamed spans will be finished separately once the stream generator is exhausted
+        if span.error or not stream:
+            span.finish()
+    return generations
+
+
+@with_traced_module
+async def traced_agenerate(genai, pin, func, instance, args, kwargs):
+    integration = genai._datadog_integration
+    stream = kwargs.get("stream", False)
+    generations = None
+    span = integration.trace(
+        pin,
+        "%s.%s" % (instance.__class__.__name__, func.__name__),
+        provider="google",
+    )
+    try:
+        tag_request(span, integration, instance, args, kwargs)
+        generations = await func(*args, **kwargs)
+        if stream:
+            return TracedAsyncGenerateContentResponse(generations, instance, integration, span, args, kwargs)
+        tag_response(span, generations, integration, instance)
+    except Exception:
+        span.set_exc_info(*sys.exc_info())
+        raise
+    finally:
+        # streamed spans will be finished separately once the stream generator is exhausted
+        if span.error or not stream:
+            span.finish()
+    return generations
+
+
+def patch():
+    if getattr(genai, "_datadog_patch", False):
+        return
+
+    genai._datadog_patch = True
+
+    Pin().onto(genai)
```

Comment:
Just a sanity check on my end: Are we using `Pin` here because it's common to have multiple processes/contexts where an instance of `genai` gets accessed?

Reply:
Kind of, this way we can use the same tracer instance - this is just standard practice following the other integrations. Do you think we should avoid using Pin here?

Reply:
No not necessarily. I just had a conversation with someone a while ago that Pin is not always necessary, so I always tend to dig into how Pin gets used in new integrations. Either way, now that we abstracted implementation details in `internal/` it's not a blocker, just my own curiosity 🙂